framework/include/utils/NestedSolve.h

```diff
@@ -0,0 +1,336 @@
+                                                          NestedSolve::DynamicVector,
+                                                          Eigen::Matrix<Real, N, 1>>::type>::type;
+
+  /// Jacobian matrix type
+  template <int N = 0>
+  using Jacobian =
+      typename std::conditional<N == 1,
+                                Real,
+                                typename std::conditional<N == 0,
+                                                          NestedSolve::DynamicMatrix,
+                                                          Eigen::Matrix<Real, N, N>>::type>::type;
+
+  /**
+   * Solve the N*N nonlinear equation system
+   */
+  template <typename V, typename T>
+  void nonlinear(V & guess, T compute);
+
+  template <typename R, typename J>
+  void nonlinear(DynamicVector & guess, R computeResidual, J computeJacobian);
+  template <typename R, typename J>
+  void nonlinear(Real & guess, R computeResidual, J computeJacobian);
+  template <typename R, typename J>
+  void nonlinear(RealVectorValue & guess, R computeResidual, J computeJacobian);
+
+  ///@{ default values
+  static Real relativeToleranceDefault() { return 1e-8; }
+  static Real absoluteToleranceDefault() { return 1e-13; }
+  static unsigned int minIterationsDefault() { return 10; }
+  static unsigned int maxIterationsDefault() { return 1000; }
+  ///@}
+
+  void setRelativeTolerance(Real rel) { _relative_tolerance_square = rel * rel; }
+  void setAbsoluteTolerance(Real abs) { _absolute_tolerance_square = abs * abs; }
+
+  Real _relative_tolerance_square;
+  Real _absolute_tolerance_square;
+  unsigned int _min_iterations;
+  unsigned int _max_iterations;
+
+  enum class State
+  {
+    NONE,
+    CONVERGED_ABS,
+    CONVERGED_REL,
+    EXACT_GUESS
+  };
+
+protected:
+  /// current solver state
+  State _state;
+
+  ///@{ Deduce the Jacobian type from the solution type
+  template <typename T>
+  struct CorrespondingJacobianTempl;
+
+  template <int N>
+  struct CorrespondingJacobianTempl<Eigen::Matrix<Real, N, 1>>
+  {
+    using type = Eigen::Matrix<Real, N, N>;
+  };
+
+  template <typename T>
+  using CorrespondingJacobian = typename CorrespondingJacobianTempl<T>::type;
+  ///@}
+
+  /**
+   * Size a dynamic Jacobian matrix correctly
+   */
+  void sizeItems(const NestedSolve::DynamicVector & guess,
+                 NestedSolve::DynamicVector & residual,
+                 NestedSolve::DynamicMatrix & jacobian) const
+  {
+    const auto N = guess.size();
+    residual.resize(N, 1);
+    jacobian.resize(N, N);
+  }
+
+  /**
+   * Sizing is a no-op for compile time sized types (and scalars)
+   */
+  template <typename V, typename T>
+  void sizeItems(const V &, V &, T &) const
+  {
+  }
+
+  /**
+   *  Solve A*x=b for x
+   */
+  template <typename J, typename V>
+  void linear(const J & A, V & x, const V & b) const
+  {
+    // we could make the linear solve method configurable here
+    x = A.colPivHouseholderQr().solve(b);
```

Comment:
Yeah, maybe an option would be good. I was going off of this https://eigen.tuxfamily.org/dox/group__TutorialLinearAlgebra.html page and chose the solver that has no restrictions on the matrix and the highest accuracy. An option for this could be added in a follow-on PR.

Reply:
It makes sense.